File: examples/python/quickstart.py
#!/usr/bin/env python3
"""
VCP Quick Start - Experience "Verify, Don't Trust" in 5 minutes

Usage:
    export VCP_API_KEY="your-api-key"
    pip install httpx
    python quickstart.py

Get your API key at: https://veritaschain.org/register
"""
import os
import sys
import hashlib

try:
    import httpx
except ImportError:
    print("Error: httpx is required. Install with: pip install httpx")
    sys.exit(1)

# Configuration
API_BASE = "https://explorer.veritaschain.org/api/v1"
API_KEY = os.environ.get("VCP_API_KEY", "")

if not API_KEY:
    print("Warning: VCP_API_KEY not set. Some features may not work.")
    print("Get your API key at: https://veritaschain.org/register")
    print()

headers = {"Authorization": f"Bearer {API_KEY}"} if API_KEY else {}


def verify_merkle_proof(event_hash: str, proof_path: list, merkle_root: str) -> bool:
    """
    Verify Merkle proof locally (RFC 6962 compliant).
    
    This function runs entirely on your machine.
    No server trust required - that's the "Verify, Don't Trust" principle!
    
    Args:
        event_hash: SHA-256 hash of the event (hex string)
        proof_path: List of sibling hashes with positions
        merkle_root: Expected Merkle root (hex string)
    
    Returns:
        True if proof is valid, False otherwise
    """
    try:
        current = bytes.fromhex(event_hash)
        
        for step in proof_path:
            sibling = bytes.fromhex(step["hash"])
            if step["position"] == "left":
                # Sibling is on the left, so: H(sibling || current)
                combined = sibling + current
            else:
                # Sibling is on the right, so: H(current || sibling)
                combined = current + sibling
            current = hashlib.sha256(combined).digest()
        
        return current.hex() == merkle_root
    except Exception as e:
        print(f"Verification error: {e}")
        return False


def main():
    print("=" * 60)
    print("VCP Quick Start")
    print("VeritasChain Protocol - \"Verify, Don't Trust\"")
    print("=" * 60)
    
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    # 1. System Status (No authentication required)
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    print("\n📊 1. System Status")
    print("-" * 40)
    
    try:
        response = httpx.get(f"{API_BASE}/system/status", timeout=10.0)
        response.raise_for_status()
        status = response.json()
        
        total_events = status.get('total_events', 'N/A')
        if isinstance(total_events, int):
            total_events = f"{total_events:,}"
        
        print(f"Total Events:  {total_events}")
        print(f"VCP Version:   {status.get('vcp_version', '1.0')}")
        print(f"API Version:   {status.get('api_version', '1.1')}")
        
        if status.get('last_anchor'):
            anchor = status['last_anchor']
            print(f"Last Anchor:   {anchor.get('network', 'N/A')}")
            print(f"               Block #{anchor.get('block_number', 'N/A')}")
    except httpx.HTTPError as e:
        print(f"Error fetching status: {e}")
        return
    
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    # 2. Search Events (Requires API key)
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    print("\n🔍 2. Recent Events")
    print("-" * 40)
    
    if not API_KEY:
        print("Skipped: API key required")
        print("Set VCP_API_KEY environment variable to continue")
        return
    
    try:
        response = httpx.get(
            f"{API_BASE}/events",
            headers=headers,
            params={"limit": 5},
            timeout=10.0
        )
        
        if response.status_code == 401:
            print("Error: Invalid API key")
            print("Get a valid key at: https://veritaschain.org/register")
            return
        
        response.raise_for_status()
        result = response.json()
        events = result.get("events", [])
        
        if not events:
            print("No events found")
            print("\nThis might mean:")
            print("  - The sandbox is empty")
            print("  - Your API key doesn't have access to any events")
            return
        
        print(f"Found {len(events)} events:\n")
        for i, event in enumerate(events, 1):
            h = event["header"]
            event_type = h.get('event_type', '?')
            symbol = h.get('symbol', '?')
            event_id = h.get('event_id', '?')[:24]
            timestamp = h.get('timestamp_iso', '?')[:19]
            print(f"  [{i}] {event_type:3} | {symbol:8} | {event_id}... | {timestamp}")
        
    except httpx.HTTPError as e:
        print(f"Error fetching events: {e}")
        return
    
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    # 3. Merkle Proof Verification
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    print("\n🔐 3. Merkle Proof Verification")
    print("-" * 40)
    
    # Get proof for the first event
    event_id = events[0]["header"]["event_id"]
    
    try:
        response = httpx.get(
            f"{API_BASE}/events/{event_id}/proof",
            headers=headers,
            timeout=10.0
        )
        
        if response.status_code == 404:
            print(f"Proof not available for event: {event_id[:32]}...")
            print("The event may be too recent (not yet anchored)")
            return
        
        response.raise_for_status()
        proof = response.json()
        
        print(f"Event ID:     {event_id[:32]}...")
        print(f"Event Hash:   {proof['event_hash'][:32]}...")
        print(f"Merkle Root:  {proof['merkle_root'][:32]}...")
        
        tree_size = proof.get('tree_size', 'N/A')
        if isinstance(tree_size, int):
            tree_size = f"{tree_size:,}"
        print(f"Tree Size:    {tree_size}")
        print(f"Leaf Index:   {proof.get('leaf_index', 'N/A')}")
        print(f"Proof Steps:  {len(proof.get('proof_path', []))}")
        
        # Perform local verification
        print("\nVerifying locally...")
        is_valid = verify_merkle_proof(
            proof["event_hash"],
            proof.get("proof_path", []),
            proof["merkle_root"]
        )
        
        print()
        if is_valid:
            print("✅ VERIFIED: Proof is mathematically valid!")
            print()
            print("   What this means:")
            print("   → The event exists in the Merkle tree")
            print("   → The proof was verified on YOUR machine")
            print("   → No server trust was required")
            print("   → Any tampering would have been detected")
        else:
            print("❌ FAILED: Proof verification failed!")
            print()
            print("   This could mean:")
            print("   → Data was tampered with")
            print("   → Proof is incomplete or corrupted")
            print("   → There's a bug in the verification code")
            
    except httpx.HTTPError as e:
        print(f"Error fetching proof: {e}")
        return
    
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    # 4. Sample Event Structure
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    print("\n📝 4. VCP Event Structure")
    print("-" * 40)
    print("""
A VCP event consists of three parts:

┌─────────────────────────────────────────┐
│ HEADER                                  │
│  • event_id (UUID v7)                   │
│  • trace_id (links related events)      │
│  • timestamp_int (nanoseconds)          │
│  • timestamp_iso (ISO 8601)             │
│  • event_type (SIG, ORD, EXE, etc.)     │
│  • venue_id, symbol, account_id         │
├─────────────────────────────────────────┤
│ PAYLOAD                                 │
│  • vcp_trade (orders, executions)       │
│  • vcp_risk (risk snapshots)            │
│  • vcp_gov (algorithm governance)       │
├─────────────────────────────────────────┤
│ SECURITY                                │
│  • prev_hash (chain link)               │
│  • event_hash (this event's hash)       │
│  • signature (Ed25519)                  │
└─────────────────────────────────────────┘
""")
    
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    # Done
    # ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    print("=" * 60)
    print("🎉 Quick Start Complete!")
    print("=" * 60)
    print()
    print("You've just experienced the core of VCP:")
    print("  ✓ Queried the VCP Explorer API")
    print("  ✓ Retrieved a Merkle proof")
    print("  ✓ Verified it locally (no server trust!)")
    print()
    print("Next steps:")
    print("  📖 Full Spec:      https://github.com/veritaschain/vcp-spec")
    print("  🛠️  SDK Spec:       https://github.com/veritaschain/vcp-sdk-spec")
    print("  🔍 Explorer GUI:   https://veritaschain.org/explorer/app/")
    print("  🔌 MT4/MT5:        https://github.com/veritaschain/vcp-sidecar-guide")
    print("  ✅ Get Certified:  https://certified.veritaschain.org")
    print()


if __name__ == "__main__":
    main()

File: examples/typescript/quickstart.ts
/**
 * VCP Quick Start - Experience "Verify, Don't Trust" in 5 minutes
 * 
 * Usage:
 *   export VCP_API_KEY="your-api-key"
 *   npx ts-node quickstart.ts
 *   # or
 *   npx tsx quickstart.ts
 * 
 * Get your API key at: https://veritaschain.org/register
 */

import crypto from 'crypto';

// Configuration
const API_BASE = 'https://explorer.veritaschain.org/api/v1';
const API_KEY = process.env.VCP_API_KEY || '';

if (!API_KEY) {
  console.log('Warning: VCP_API_KEY not set. Some features may not work.');
  console.log('Get your API key at: https://veritaschain.org/register\n');
}

// Types
interface ProofStep {
  hash: string;
  position: 'left' | 'right';
}

interface MerkleProof {
  event_hash: string;
  merkle_root: string;
  proof_path: ProofStep[];
  tree_size: number;
  leaf_index: number;
}

interface VcpEvent {
  header: {
    event_id: string;
    trace_id: string;
    timestamp_int: string;
    timestamp_iso: string;
    event_type: string;
    event_type_code: number;
    venue_id: string;
    symbol: string;
    account_id: string;
  };
  payload: Record<string, unknown>;
  security: {
    prev_hash: string;
    event_hash: string;
    hash_algo: string;
    signature?: string;
  };
}

/**
 * Verify Merkle proof locally (RFC 6962 compliant).
 * 
 * This function runs entirely on your machine.
 * No server trust required - that's the "Verify, Don't Trust" principle!
 */
function verifyMerkleProof(proof: MerkleProof): boolean {
  try {
    let current = Buffer.from(proof.event_hash, 'hex');
    
    for (const step of proof.proof_path) {
      const sibling = Buffer.from(step.hash, 'hex');
      const combined = step.position === 'left'
        ? Buffer.concat([sibling, current])  // H(sibling || current)
        : Buffer.concat([current, sibling]); // H(current || sibling)
      current = crypto.createHash('sha256').update(combined).digest();
    }
    
    return current.toString('hex') === proof.merkle_root;
  } catch (error) {
    console.error('Verification error:', error);
    return false;
  }
}

/**
 * Helper to make authenticated requests
 */
async function apiRequest<T>(endpoint: string): Promise<T> {
  const headers: Record<string, string> = {};
  if (API_KEY) {
    headers['Authorization'] = `Bearer ${API_KEY}`;
  }
  
  const response = await fetch(`${API_BASE}${endpoint}`, { headers });
  
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }
  
  return response.json();
}

async function main() {
  console.log('='.repeat(60));
  console.log('VCP Quick Start');
  console.log('VeritasChain Protocol - "Verify, Don\'t Trust"');
  console.log('='.repeat(60));
  
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 1. System Status (No authentication required)
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  console.log('\n📊 1. System Status');
  console.log('-'.repeat(40));
  
  try {
    const status = await fetch(`${API_BASE}/system/status`).then(r => r.json());
    
    const totalEvents = status.total_events?.toLocaleString() ?? 'N/A';
    console.log(`Total Events:  ${totalEvents}`);
    console.log(`VCP Version:   ${status.vcp_version ?? '1.0'}`);
    console.log(`API Version:   ${status.api_version ?? '1.1'}`);
    
    if (status.last_anchor) {
      console.log(`Last Anchor:   ${status.last_anchor.network ?? 'N/A'}`);
      console.log(`               Block #${status.last_anchor.block_number ?? 'N/A'}`);
    }
  } catch (error) {
    console.error('Error fetching status:', error);
    return;
  }
  
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 2. Search Events (Requires API key)
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  console.log('\n🔍 2. Recent Events');
  console.log('-'.repeat(40));
  
  if (!API_KEY) {
    console.log('Skipped: API key required');
    console.log('Set VCP_API_KEY environment variable to continue');
    return;
  }
  
  let events: VcpEvent[] = [];
  
  try {
    const result = await apiRequest<{ events: VcpEvent[]; total_count: number }>(
      '/events?limit=5'
    );
    events = result.events || [];
    
    if (events.length === 0) {
      console.log('No events found');
      return;
    }
    
    console.log(`Found ${events.length} events:\n`);
    events.forEach((event, i) => {
      const h = event.header;
      const timestamp = h.timestamp_iso?.slice(0, 19) ?? '?';
      console.log(
        `  [${i + 1}] ${h.event_type.padEnd(3)} | ` +
        `${h.symbol.padEnd(8)} | ` +
        `${h.event_id.slice(0, 24)}... | ` +
        `${timestamp}`
      );
    });
  } catch (error) {
    console.error('Error fetching events:', error);
    return;
  }
  
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // 3. Merkle Proof Verification
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  console.log('\n🔐 3. Merkle Proof Verification');
  console.log('-'.repeat(40));
  
  const eventId = events[0].header.event_id;
  
  try {
    const proof = await apiRequest<MerkleProof>(`/events/${eventId}/proof`);
    
    console.log(`Event ID:     ${eventId.slice(0, 32)}...`);
    console.log(`Event Hash:   ${proof.event_hash.slice(0, 32)}...`);
    console.log(`Merkle Root:  ${proof.merkle_root.slice(0, 32)}...`);
    console.log(`Tree Size:    ${proof.tree_size?.toLocaleString() ?? 'N/A'}`);
    console.log(`Leaf Index:   ${proof.leaf_index ?? 'N/A'}`);
    console.log(`Proof Steps:  ${proof.proof_path?.length ?? 0}`);
    
    // Perform local verification
    console.log('\nVerifying locally...');
    const isValid = verifyMerkleProof(proof);
    
    console.log();
    if (isValid) {
      console.log('✅ VERIFIED: Proof is mathematically valid!');
      console.log();
      console.log('   What this means:');
      console.log('   → The event exists in the Merkle tree');
      console.log('   → The proof was verified on YOUR machine');
      console.log('   → No server trust was required');
      console.log('   → Any tampering would have been detected');
    } else {
      console.log('❌ FAILED: Proof verification failed!');
      console.log();
      console.log('   This could mean:');
      console.log('   → Data was tampered with');
      console.log('   → Proof is incomplete or corrupted');
    }
  } catch (error) {
    console.error('Error fetching proof:', error);
    return;
  }
  
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  // Done
  // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  console.log('\n' + '='.repeat(60));
  console.log('🎉 Quick Start Complete!');
  console.log('='.repeat(60));
  console.log();
  console.log('You\'ve just experienced the core of VCP:');
  console.log('  ✓ Queried the VCP Explorer API');
  console.log('  ✓ Retrieved a Merkle proof');
  console.log('  ✓ Verified it locally (no server trust!)');
  console.log();
  console.log('Next steps:');
  console.log('  📖 Full Spec:      https://github.com/veritaschain/vcp-spec');
  console.log('  🛠️  SDK Spec:       https://github.com/veritaschain/vcp-sdk-spec');
  console.log('  🔍 Explorer GUI:   https://veritaschain.org/explorer/app/');
  console.log('  🔌 MT4/MT5:        https://github.com/veritaschain/vcp-sidecar-guide');
  console.log('  ✅ Get Certified:  https://certified.veritaschain.org');
  console.log();
}

main().catch(console.error);
